plugins/parsers/avro/parser.go

```diff
@@ -0,0 +1,218 @@
+		if err != nil {
+			return nil, err
+		}
+		schema = schemastruct.Schema
+		codec = schemastruct.Codec
+	} else {
+		schema = p.Schema
+		codec, err = goavro.NewCodec(schema)
+		if err != nil {
+			return nil, err
+		}
+	}
+	native, _, err := codec.NativeFromBinary(binaryData)
+	if err != nil {
+		return nil, err
+	}
+	// Cast to string-to-interface
+	codecSchema := native.(map[string]interface{})
+	m, err := p.createMetric(codecSchema, schema)
+	if err != nil {
+		return nil, err
+	}
+
+	return []telegraf.Metric{m}, nil
+}
+
+func (p *Parser) extractSchemaAndMessage(buf []byte) (int, []byte, error) {
+	if len(buf) < 5 {
+		err := fmt.Errorf("buf is %d bytes; must be at least 5", len(buf))
+		return 0, nil, err
+	}
+	schemaID := int(binary.BigEndian.Uint32(buf[1:5]))
+	binaryData := buf[5:]
+	return schemaID, binaryData, nil
+}
+
+func (p *Parser) ParseLine(line string) (telegraf.Metric, error) {
+	metrics, err := p.Parse([]byte(line))
+	if err != nil {
+		return nil, err
+	}
+
+	if len(metrics) != 1 {
+		return nil, fmt.Errorf("line contains multiple metrics")
+	}
+
+	return metrics[0], nil
+}
+
+func (p *Parser) SetDefaultTags(tags map[string]string) {
+	p.DefaultTags = tags
+}
+
+func (p *Parser) createMetric(data map[string]interface{}, schema string) (telegraf.Metric, error) {
+	now := time.Now().Unix()
+	timestamp, err := internal.ParseTimestamp(p.TimestampFormat, now, "UTC")
+	if err != nil {
+		return nil, fmt.Errorf("parsing timestamp from %q failed: %w", p.Timestamp, err)
+	}
+
+	// Tags differ from fields, in that tags are inherently strings.
+	// fields can be of any type.
+	fields := make(map[string]interface{})
+	tags := make(map[string]string)
+
+	// Set default tag values
+	for k, v := range p.DefaultTags {
+		tags[k] = v
+	}
+	// Avro doesn't have a Tag/Field distinction, so we have to tell
+	// Telegraf which items are our tags.
+	for _, tag := range p.Tags {
+		tags[tag], err = internal.ToString(data[tag])
+		if err != nil {
+			p.Log.Warnf("Could not convert %v to string for tag %q: %v", data[tag], tag, err)
+		}
+	}
+	var fieldList []string
+	if len(p.Fields) != 0 {
+		// If you have specified your fields in the config, you
+		// get what you asked for.
+		fieldList = p.Fields
+	} else {
+		for k := range data {
+			// Otherwise, that which is not a tag is a field
+			if _, ok := tags[k]; !ok {
+				fieldList = append(fieldList, k)
+			}
+		}
+	}
+	flatFields := make(map[string]interface{})
+	// We need to flatten out our fields.  The default (the separator
+	// string is empty) is equivalent to what streamreactor does.
+	sep := flatten.SeparatorStyle{
+		Before: "",
+		Middle: p.FieldSeparator,
+		After:  "",
+	}
+	for _, fld := range fieldList {
+		candidate := make(map[string]interface{})
+		candidate[fld] = data[fld] // 1-item map
+		flat, err := flatten.Flatten(candidate, "", sep)
+		if err != nil {
+			return nil, fmt.Errorf("Failed to flatten field %s: %v", fld, err)
```

Comment:
Error messages need to start lower-case. Furthermore, please always wrap errors so we can check for the error-type later on
```suggestion
			return nil, fmt.Errorf("flatten field %q failed: %w", fld, err)
```